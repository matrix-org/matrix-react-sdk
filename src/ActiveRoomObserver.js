/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import RoomViewStore from './stores/RoomViewStore';

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
class ActiveRoomObserver {
    constructor() {
        this._listeners = {};

        this._activeRoomId = RoomViewStore.getRoomId();
        // TODO: We could self-destruct when the last listener goes away, or at least
        // stop listening.
        this._roomStoreToken = RoomViewStore.addListener(this._onRoomViewStoreUpdate.bind(this));
    }

    addListener(roomId, listener) {
        if (!this._listeners[roomId]) this._listeners[roomId] = [];
        this._listeners[roomId].push(listener);
    }

    removeListener(roomId, listener) {
        if (this._listeners[roomId]) {
            const i = this._listeners[roomId].indexOf(listener);
            if (i > -1) {
                this._listeners[roomId].splice(i, 1);
            }
        } else {
            console.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
        }
    }

    _emit(roomId) {
        if (!this._listeners[roomId]) return;

        for (const l of this._listeners[roomId]) {
            l.call();
        }
    }

    _onRoomViewStoreUpdate() {
        // emit for the old room ID
        if (this._activeRoomId) this._emit(this._activeRoomId);

        // update our cache
        this._activeRoomId = RoomViewStore.getRoomId();

        // and emit for the new one
        if (this._activeRoomId) this._emit(this._activeRoomId);
    }
}

if (global.mx_ActiveRoomObserver === undefined) {
    global.mx_ActiveRoomObserver = new ActiveRoomObserver();
}
export default global.mx_ActiveRoomObserver;
