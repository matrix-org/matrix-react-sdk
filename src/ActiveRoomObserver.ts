/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import RoomViewStore from './stores/RoomViewStore';

type Listener = (isActive: boolean) => void;

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
export class ActiveRoomObserver {
    private listeners: {[key: string]: Listener[]} = {};
    private _activeRoomId = RoomViewStore.getRoomId();
    private readonly roomStoreToken: string;

    constructor() {
        // TODO: We could self-destruct when the last listener goes away, or at least stop listening.
        this.roomStoreToken = RoomViewStore.addListener(this.onRoomViewStoreUpdate);
    }

    public get activeRoomId(): string {
        return this._activeRoomId;
    }

    public addListener(roomId, listener) {
        if (!this.listeners[roomId]) this.listeners[roomId] = [];
        this.listeners[roomId].push(listener);
    }

    public removeListener(roomId, listener) {
        if (this.listeners[roomId]) {
            const i = this.listeners[roomId].indexOf(listener);
            if (i > -1) {
                this.listeners[roomId].splice(i, 1);
            }
        } else {
            console.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
        }
    }

    private emit(roomId, isActive: boolean) {
        if (!this.listeners[roomId]) return;

        for (const l of this.listeners[roomId]) {
            l.call(null, isActive);
        }
    }

    private onRoomViewStoreUpdate = () => {
        // emit for the old room ID
        if (this._activeRoomId) this.emit(this._activeRoomId, false);

        // update our cache
        this._activeRoomId = RoomViewStore.getRoomId();

        // and emit for the new one
        if (this._activeRoomId) this.emit(this._activeRoomId, true);
    };
}

if (window.mxActiveRoomObserver === undefined) {
    window.mxActiveRoomObserver = new ActiveRoomObserver();
}
export default window.mxActiveRoomObserver;
