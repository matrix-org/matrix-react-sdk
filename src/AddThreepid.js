/*
Copyright 2016 OpenMarket Ltd
Copyright 2017 Vector Creations Ltd
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import {MatrixClientPeg} from './MatrixClientPeg';
import * as sdk from './index';
import Modal from './Modal';
import { _t } from './languageHandler';
import IdentityAuthClient from './IdentityAuthClient';

function getIdServerDomain() {
    return MatrixClientPeg.get().idBaseUrl.split("://")[1];
}

/**
 * Allows a user to add a third party identifier to their homeserver and,
 * optionally, the identity servers.
 *
 * This involves getting an email token from the identity server to "prove" that
 * the client owns the given email address, which is then passed to the
 * add threepid API on the homeserver.
 *
 * Diagrams of the intended API flows here are available at:
 *
 * https://gist.github.com/jryans/839a09bf0c5a70e2f36ed990d50ed928
 */
export default class AddThreepid {
    constructor() {
        this.clientSecret = MatrixClientPeg.get().generateClientSecret();
        this.sessionId = null;
        this.submitUrl = null;
    }

    /**
     * Attempt to add an email threepid to the homeserver.
     * This will trigger a side-effect of sending an email to the provided email address.
     * @param {string} emailAddress The email address to add
     * @return {Promise} Resolves when the email has been sent. Then call checkEmailLinkClicked().
     */
    addEmailAddress(emailAddress) {
        return MatrixClientPeg.get().requestAdd3pidEmailToken(emailAddress, this.clientSecret, 1).then((res) => {
            this.sessionId = res.sid;
            return res;
        }, function(err) {
            if (err.errcode === 'M_THREEPID_IN_USE') {
                err.message = _t('This email address is already in use');
            } else if (err.httpStatus) {
                err.message = err.message + ` (Status ${err.httpStatus})`;
            }
            throw err;
        });
    }

    /**
     * Attempt to bind an email threepid on the identity server via the homeserver.
     * This will trigger a side-effect of sending an email to the provided email address.
     * @param {string} emailAddress The email address to add
     * @return {Promise} Resolves when the email has been sent. Then call checkEmailLinkClicked().
     */
    async bindEmailAddress(emailAddress) {
        this.bind = true;
        if (await MatrixClientPeg.get().doesServerSupportSeparateAddAndBind()) {
            // For separate bind, request a token directly from the IS.
            const authClient = new IdentityAuthClient();
            const identityAccessToken = await authClient.getAccessToken();
            return MatrixClientPeg.get().requestEmailToken(
                emailAddress, this.clientSecret, 1,
                undefined, undefined, identityAccessToken,
            ).then((res) => {
                this.sessionId = res.sid;
                return res;
            }, function(err) {
                if (err.errcode === 'M_THREEPID_IN_USE') {
                    err.message = _t('This email address is already in use');
                } else if (err.httpStatus) {
                    err.message = err.message + ` (Status ${err.httpStatus})`;
                }
                throw err;
            });
        } else {
            // For tangled bind, request a token via the HS.
            return this.addEmailAddress(emailAddress);
        }
    }

    /**
     * Attempt to add a MSISDN threepid to the homeserver.
     * This will trigger a side-effect of sending an SMS to the provided phone number.
     * @param {string} phoneCountry The ISO 2 letter code of the country to resolve phoneNumber in
     * @param {string} phoneNumber The national or international formatted phone number to add
     * @return {Promise} Resolves when the text message has been sent. Then call haveMsisdnToken().
     */
    addMsisdn(phoneCountry, phoneNumber) {
        return MatrixClientPeg.get().requestAdd3pidMsisdnToken(
            phoneCountry, phoneNumber, this.clientSecret, 1,
        ).then((res) => {
            this.sessionId = res.sid;
            this.submitUrl = res.submit_url;
            return res;
        }, function(err) {
            if (err.errcode === 'M_THREEPID_IN_USE') {
                err.message = _t('This phone number is already in use');
            } else if (err.httpStatus) {
                err.message = err.message + ` (Status ${err.httpStatus})`;
            }
            throw err;
        });
    }

    /**
     * Attempt to bind a MSISDN threepid on the identity server via the homeserver.
     * This will trigger a side-effect of sending an SMS to the provided phone number.
     * @param {string} phoneCountry The ISO 2 letter code of the country to resolve phoneNumber in
     * @param {string} phoneNumber The national or international formatted phone number to add
     * @return {Promise} Resolves when the text message has been sent. Then call haveMsisdnToken().
     */
    async bindMsisdn(phoneCountry, phoneNumber) {
        this.bind = true;
        if (await MatrixClientPeg.get().doesServerSupportSeparateAddAndBind()) {
            // For separate bind, request a token directly from the IS.
            const authClient = new IdentityAuthClient();
            const identityAccessToken = await authClient.getAccessToken();
            return MatrixClientPeg.get().requestMsisdnToken(
                phoneCountry, phoneNumber, this.clientSecret, 1,
                undefined, undefined, identityAccessToken,
            ).then((res) => {
                this.sessionId = res.sid;
                return res;
            }, function(err) {
                if (err.errcode === 'M_THREEPID_IN_USE') {
                    err.message = _t('This phone number is already in use');
                } else if (err.httpStatus) {
                    err.message = err.message + ` (Status ${err.httpStatus})`;
                }
                throw err;
            });
        } else {
            // For tangled bind, request a token via the HS.
            return this.addMsisdn(phoneCountry, phoneNumber);
        }
    }

    /**
     * Checks if the email link has been clicked by attempting to add the threepid
     * @return {Promise} Resolves if the email address was added. Rejects with an object
     * with a "message" property which contains a human-readable message detailing why
     * the request failed.
     */
    async checkEmailLinkClicked() {
        try {
            if (await MatrixClientPeg.get().doesServerSupportSeparateAddAndBind()) {
                if (this.bind) {
                    const authClient = new IdentityAuthClient();
                    const identityAccessToken = await authClient.getAccessToken();
                    await MatrixClientPeg.get().bindThreePid({
                        sid: this.sessionId,
                        client_secret: this.clientSecret,
                        id_server: getIdServerDomain(),
                        id_access_token: identityAccessToken,
                    });
                } else {
                    try {
                        await this._makeAddThreepidOnlyRequest();

                        // The spec has always required this to use UI auth but synapse briefly
                        // implemented it without, so this may just succeed and that's OK.
                        return;
                    } catch (e) {
                        if (e.httpStatus !== 401 || !e.data || !e.data.flows) {
                            // doesn't look like an interactive-auth failure
                            throw e;
                        }

                        // pop up an interactive auth dialog
                        const InteractiveAuthDialog = sdk.getComponent("dialogs.InteractiveAuthDialog");

                        const { finished } = Modal.createTrackedDialog('Add Email', '', InteractiveAuthDialog, {
                            title: _t("Add Email Address"),
                            matrixClient: MatrixClientPeg.get(),
                            authData: e.data,
                            makeRequest: this._makeAddThreepidOnlyRequest,
                        });
                        return finished;
                    }
                }
            } else {
                await MatrixClientPeg.get().addThreePid({
                    sid: this.sessionId,
                    client_secret: this.clientSecret,
                    id_server: getIdServerDomain(),
                }, this.bind);
            }
        } catch (err) {
            if (err.httpStatus === 401) {
                err.message = _t('Failed to verify email address: make sure you clicked the link in the email');
            } else if (err.httpStatus) {
                err.message += ` (Status ${err.httpStatus})`;
            }
            throw err;
        }
    }

    /**
     * @param {Object} auth UI auth object
     * @return {Promise<Object>} Response from /3pid/add call (in current spec, an empty object)
     */
    _makeAddThreepidOnlyRequest = (auth) => {
        return MatrixClientPeg.get().addThreePidOnly({
            sid: this.sessionId,
            client_secret: this.clientSecret,
            auth,
        });
    }

    /**
     * Takes a phone number verification code as entered by the user and validates
     * it with the ID server, then if successful, adds the phone number.
     * @param {string} msisdnToken phone number verification code as entered by the user
     * @return {Promise} Resolves if the phone number was added. Rejects with an object
     * with a "message" property which contains a human-readable message detailing why
     * the request failed.
     */
    async haveMsisdnToken(msisdnToken) {
        const authClient = new IdentityAuthClient();
        const supportsSeparateAddAndBind =
            await MatrixClientPeg.get().doesServerSupportSeparateAddAndBind();

        let result;
        if (this.submitUrl) {
            result = await MatrixClientPeg.get().submitMsisdnTokenOtherUrl(
                this.submitUrl,
                this.sessionId,
                this.clientSecret,
                msisdnToken,
            );
        } else if (this.bind || !supportsSeparateAddAndBind) {
            result = await MatrixClientPeg.get().submitMsisdnToken(
                this.sessionId,
                this.clientSecret,
                msisdnToken,
                await authClient.getAccessToken(),
            );
        } else {
            throw new Error("The add / bind with MSISDN flow is misconfigured");
        }
        if (result.errcode) {
            throw result;
        }

        if (supportsSeparateAddAndBind) {
            if (this.bind) {
                await MatrixClientPeg.get().bindThreePid({
                    sid: this.sessionId,
                    client_secret: this.clientSecret,
                    id_server: getIdServerDomain(),
                    id_access_token: await authClient.getAccessToken(),
                });
            } else {
                try {
                    await this._makeAddThreepidOnlyRequest();

                    // The spec has always required this to use UI auth but synapse briefly
                    // implemented it without, so this may just succeed and that's OK.
                    return;
                } catch (e) {
                    if (e.httpStatus !== 401 || !e.data || !e.data.flows) {
                        // doesn't look like an interactive-auth failure
                        throw e;
                    }

                    // pop up an interactive auth dialog
                    const InteractiveAuthDialog = sdk.getComponent("dialogs.InteractiveAuthDialog");

                    const { finished } = Modal.createTrackedDialog('Add MSISDN', '', InteractiveAuthDialog, {
                        title: _t("Add Phone Number"),
                        matrixClient: MatrixClientPeg.get(),
                        authData: e.data,
                        makeRequest: this._makeAddThreepidOnlyRequest,
                    });
                    return finished;
                }
            }
        } else {
            await MatrixClientPeg.get().addThreePid({
                sid: this.sessionId,
                client_secret: this.clientSecret,
                id_server: getIdServerDomain(),
            }, this.bind);
        }
    }
}
