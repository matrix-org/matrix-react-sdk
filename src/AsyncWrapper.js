/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import * as sdk from './index';
import PropTypes from 'prop-types';
import { _t } from './languageHandler';

/**
 * Wrap an asynchronous loader function with a react component which shows a
 * spinner until the real component loads.
 */
export default class AsyncWrapper extends React.Component {
    static propTypes = {
        /** A promise which resolves with the real component
         */
        prom: PropTypes.object.isRequired,
    };

    state = {
        component: null,
        error: null,
    };

    componentDidMount() {
        this._unmounted = false;
        // XXX: temporary logging to try to diagnose
        // https://github.com/vector-im/element-web/issues/3148
        console.log('Starting load of AsyncWrapper for modal');
        this.props.prom.then((result) => {
            if (this._unmounted) {
                return;
            }
            // Take the 'default' member if it's there, then we support
            // passing in just an import()ed module, since ES6 async import
            // always returns a module *namespace*.
            const component = result.default ? result.default : result;
            this.setState({component});
        }).catch((e) => {
            console.warn('AsyncWrapper promise failed', e);
            this.setState({error: e});
        });
    }

    componentWillUnmount() {
        this._unmounted = true;
    }

    _onWrapperCancelClick = () => {
        this.props.onFinished(false);
    };

    render() {
        if (this.state.component) {
            const Component = this.state.component;
            return <Component {...this.props} />;
        } else if (this.state.error) {
            const BaseDialog = sdk.getComponent('views.dialogs.BaseDialog');
            const DialogButtons = sdk.getComponent('views.elements.DialogButtons');
            return <BaseDialog onFinished={this.props.onFinished}
                title={_t("Error")}
            >
                {_t("Unable to load! Check your network connectivity and try again.")}
                <DialogButtons primaryButton={_t("Dismiss")}
                    onPrimaryButtonClick={this._onWrapperCancelClick}
                    hasCancel={false}
                />
            </BaseDialog>;
        } else {
            // show a spinner until the component is loaded.
            const Spinner = sdk.getComponent("elements.Spinner");
            return <Spinner />;
        }
    }
}

