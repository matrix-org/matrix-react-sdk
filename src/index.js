/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
