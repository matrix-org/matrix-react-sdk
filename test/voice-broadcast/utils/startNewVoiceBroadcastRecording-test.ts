/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { mocked } from "jest-mock";
import { EventType, MatrixClient, MatrixEvent, Room, RoomStateEvent } from "matrix-js-sdk/src/matrix";

import {
    startNewVoiceBroadcastRecording,
    VoiceBroadcastInfoEventType,
    VoiceBroadcastInfoState,
    VoiceBroadcastRecordingsStore,
    VoiceBroadcastRecording,
} from "../../../src/voice-broadcast";
import { mkEvent, stubClient } from "../../test-utils";

jest.mock("../../../src/voice-broadcast/models/VoiceBroadcastRecording", () => ({
    VoiceBroadcastRecording: jest.fn(),
}));

describe("startNewVoiceBroadcastRecording", () => {
    const roomId = "!room:example.com";
    let client: MatrixClient;
    let recordingsStore: VoiceBroadcastRecordingsStore;
    let room: Room;
    let roomOnStateEventsCallbackRegistered: Promise<void>;
    let roomOnStateEventsCallbackRegisteredResolver: Function;
    let roomOnStateEventsCallback: () => void;
    let infoEvent: MatrixEvent;
    let otherEvent: MatrixEvent;
    let stateEvent: MatrixEvent;

    beforeEach(() => {
        roomOnStateEventsCallbackRegistered = new Promise((resolve) => {
            roomOnStateEventsCallbackRegisteredResolver = resolve;
        });

        room = {
            currentState: {
                getStateEvents: jest.fn().mockImplementation((type, userId) => {
                    if (type === VoiceBroadcastInfoEventType && userId === client.getUserId()) {
                        return stateEvent;
                    }
                }),
            },
            on: jest.fn().mockImplementation((eventType, callback) => {
                if (eventType === RoomStateEvent.Events) {
                    roomOnStateEventsCallback = callback;
                    roomOnStateEventsCallbackRegisteredResolver();
                }
            }),
            off: jest.fn(),
        } as unknown as Room;

        client = stubClient();
        mocked(client.getRoom).mockImplementation((getRoomId: string) => {
            if (getRoomId === roomId) {
                return room;
            }
        });
        mocked(client.sendStateEvent).mockImplementation((
            sendRoomId: string,
            eventType: string,
            _content: any,
            _stateKey: string,
        ) => {
            if (sendRoomId === roomId && eventType === VoiceBroadcastInfoEventType) {
                return Promise.resolve({ event_id: infoEvent.getId() });
            }
        });

        recordingsStore = {
            setCurrent: jest.fn(),
        } as unknown as VoiceBroadcastRecordingsStore;

        infoEvent = mkEvent({
            event: true,
            type: VoiceBroadcastInfoEventType,
            content: {
                state: VoiceBroadcastInfoState.Started,
            },
            user: client.getUserId(),
            room: roomId,
        });
        otherEvent = mkEvent({
            event: true,
            type: EventType.RoomMember,
            content: {},
            user: client.getUserId(),
            room: roomId,
        });

        mocked(VoiceBroadcastRecording).mockImplementation((
            infoEvent: MatrixEvent,
            client: MatrixClient,
        ) => {
            return {
                infoEvent,
                client,
            } as unknown as VoiceBroadcastRecording;
        });
    });

    it("should create a new Voice Broadcast", (done) => {
        let ok = false;

        startNewVoiceBroadcastRecording(roomId, client, recordingsStore).then((recording) => {
            expect(ok).toBe(true);
            expect(mocked(room.off)).toHaveBeenCalledWith(RoomStateEvent.Events, roomOnStateEventsCallback);
            expect(recording.infoEvent).toBe(infoEvent);
            done();
        });

        roomOnStateEventsCallbackRegistered.then(() => {
            // no state event, yet
            roomOnStateEventsCallback();

            // other state event
            stateEvent = otherEvent;
            roomOnStateEventsCallback();

            // the expected Voice Broadcast Info event
            stateEvent = infoEvent;
            ok = true;
            roomOnStateEventsCallback();
        });
    });
});
