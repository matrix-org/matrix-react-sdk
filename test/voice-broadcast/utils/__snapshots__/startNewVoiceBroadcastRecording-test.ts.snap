// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`startNewVoiceBroadcastRecording when the current user is allowed to send voice broadcast info state events when there already is a live broadcast of another user should show an info dialog 1`] = `
[MockFunction] {
  "calls": [
    [
      [Function],
      {
        "description": <p>
          Someone else is already recording a voice broadcast. Wait for their voice broadcast to end to start a new one.
        </p>,
        "hasCloseButton": true,
        "title": "Can't start a new voice broadcast",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`startNewVoiceBroadcastRecording when the current user is allowed to send voice broadcast info state events when there already is a live broadcast of the current user in the room should show an info dialog 1`] = `
[MockFunction] {
  "calls": [
    [
      [Function],
      {
        "description": <p>
          You are already recording a voice broadcast. Please end your current voice broadcast to start a new one.
        </p>,
        "hasCloseButton": true,
        "title": "Can't start a new voice broadcast",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`startNewVoiceBroadcastRecording when the current user is allowed to send voice broadcast info state events when there is already a current voice broadcast should show an info dialog 1`] = `
[MockFunction] {
  "calls": [
    [
      [Function],
      {
        "description": <p>
          You are already recording a voice broadcast. Please end your current voice broadcast to start a new one.
        </p>,
        "hasCloseButton": true,
        "title": "Can't start a new voice broadcast",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`startNewVoiceBroadcastRecording when the current user is not allowed to send voice broadcast info state events should show an info dialog 1`] = `
[MockFunction] {
  "calls": [
    [
      [Function],
      {
        "description": <p>
          You don't have the required permissions to start a voice broadcast in this room. Contact a room administrator to upgrade your permissions.
        </p>,
        "hasCloseButton": true,
        "title": "Can't start a new voice broadcast",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
